File: string/re125.java
package string;

class Solution {
    public boolean isPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;
        char a, b;

        
        while (i < j) {
            a = Character.toLowerCase(s.charAt(i));
            b = Character.toLowerCase(s.charAt(j));
            if (Character.isLetterOrDigit(a) && Character.isLetterOrDigit(b)) {
                if (a != b) {
                    return false;
                }
                else {
                    i++;
                    j++;
                    continue;
                }
            }
            i += (Character.isLetterOrDigit(a)) ? 0 : 1;
            j += (Character.isLetterOrDigit(b)) ? 0 : 1;
        }
        return true;
    }


    public static void main(String[] args) {
        Solution solution = new Solution();
        String s = "A man, a plan, a canal: Panama";
        System.out.println(solution.isPalindrome(s));
    }

}

File: string/re125.py
import re
"""
Runtime 45ms Beats 56.79% of users with Python3
Memory 18.01 MB Beats 20.96% of users with Python3
"""
# pdf sol is more simple.
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = re.sub('[^a-zA-Z0-9]', '', s).lower()
        # its work but SyntaxWarning: invalid escape sequence '\W'
        # s = s.lower()
        # s = re.sub("\W", '', s)
        print(s)
        return s == s[::-1]
    

# leetcode sol, two lines
# Regular expressions are not necessary for this problem
class Solution1:
    def isPalindrome(self, s: str) -> bool:
        s = [i for i in s.lower() if s.isalnum()]
        return s == s[::-1]
    

# leetcode sol, O(1) space two-pointer
class Solution2:
    def isPalindrome(self, s: str) -> bool:
        i, j = 0, len(s) - 1
        while i < j:
            a, b = s[i].lower(), s[j].lower()
            if a.isalnum() and b.isalnum():
                if a != b:
                    return False
                else:
                    i += 1
                    j += 1
                    continue
            
            i, j = i + (not a.isalnum()), j - (not b.isalnum())
        
        return True
        


s = "A man, a plan, a canal: Panama"
print(Solution().isPalindrome(s))

